designs/defaults-and-model-evolution.md

```diff
@@ -0,0 +1,770 @@
+be compatible with the shape targeted by the member and any applied constraint
+traits (for example, values for numeric types MUST be numbers that fit within
+the targeted type and match any `@length` constraints, string types match any
+`@pattern` or `@length` traits, etc).
+
+The following shapes have restrictions on their default values:
+
+* enum: can be set to any valid string _value_ of the enum.
+* intEnum: can be set to any valid integer _value_ of the enum.
+* document: can be set to `true`, `false`, string, empty list, or empty map.
+* list/set: can only be set to an empty list.
+* map: can only be set to an empty map.
+* structure: no default value.
+* union: no default value.
+
+
+### Impact on API design
+
+The `@default` trait SHOULD NOT be used for partial updates or patch style
+operations where it is critical to differentiate between omitted values and
+explicitly set values. Assigning default values is typically something that
+occurs during deserialization, and as such, it is impossible for a server to
+differentiate between whether a property was set to its default value or if a
+property was omitted.
+
+To help guide API design, built-in validation will be added to Smithy to detect
+and warn when `@default` members are detected in the top-level input of
+operations that start with `Update`, operation bound to the `update` lifecycle
+of a resource, or operations that use the `PATCH` HTTP method.
+
+For example, the following model:
+
+```
+$version: "2"
+namespace smithy.examnple
+
+operation UpdateUser {
+    input: UpdateUserInput
+}
+
+structure UpdateUserInput {
+    username: String = ""
+}
+```
+
+Would emit a warning similar to the following:
+
+```
+WARNING: smithy.example#UpdateUserInput$userName (DefaultValueInUpdate)
+     @ /path/to/main.smithy
+     |
+ 4   | operation UpdateUser {
+     | ^
+     = This update style operation has top-level input members marked with the
+       @default trait. It will be impossible to tell if the member was omitted
+       or explicitly provided. Affected members: [UpdateUserInput$username].
+```
+
+
+### Updating default values
+
+The default value of a member SHOULD NOT be changed. However, it MAY be
+necessary in extreme cases to change a default value if changing the default
+value addresses a customer-impacting issue or availability issue for a service.
+Changing default values can result in parties disagreeing on the default value
+of a member because they are using different versions of the same model.
+
+
+### Default value serialization and deserialization
+
+When deserializing a structure, a reader MUST set the member to its default
+value if the member is missing. Outside of deserialization, there MUST NOT be
+any discernible difference between an explicitly serialized member or a member
+that was assigned the default value. If such a distinction is needed, then the
+`@default` trait is inappropriate.
```

Comment:
You say that it should only be important for deser here, but a few sections down you talk about how it's needed for ser too.

Reply:
Yeah fair. I'll clean this up and simplify. I'll remove the `@default` + `@required` combo for now too so it's simpler language. We don't have a use case for that yet.